File: Trees/235. Lowest Common Ancestor of a Binary Search Tree/Solution.java
import java.util.*;

//Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}


public class Solution {
    private TreeNode dfs(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) return null;
        
        if (root.val < p.val && root.val < q.val) {
            return dfs(root.right, p, q);
        } else if (root.val > p.val && root.val > q.val) {
            return dfs(root.left, p, q);
        } else {
            return root;
        }
    }
    
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        // N : no of nodes in the Tree
        // Time Complexity: O(N)
        // Space Complexity: O(N) -> Stack Space
        // return dfs(root, p, q);


        // N : no of nodes in the Tree
        // Time Complexity: O(N)
        // Space Complexity: O(N)
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode node = stack.pop();
            if (node.val < p.val && node.val < q.val) {
                stack.push(node.right);
            } else if (node.val > p.val && node.val > q.val) {
                stack.push(node.left);
            } else {
                return node;
            }
        }
        return null;
    }
}

File: Trees/235. Lowest Common Ancestor of a Binary Search Tree/Solution.py
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        # N : no of nodes in the Tree
        # Time Complexity: O(N)
        # Space Complexity: O(N) -> Stack Space
        # def dfs(root):
        #     if not root:
        #         return None
            
        #     if root.val > p.val and root.val > q.val:
        #         return dfs(root.left)
        #     elif root.val < p.val and root.val < q.val:
        #         return dfs(root.right)
        #     else:
        #         return root
        # return dfs(root)


        # N : no of nodes in the Tree
        # Time Complexity: O(N)
        # Space Complexity: O(N)
        stack = deque([root])
        while stack:
            node = stack.pop()
            if (node.val > p.val and node.val > q.val):
                stack.append(node.left)
            elif (node.val < p.val and node.val < q.val):
                stack.append(node.right)
            else:
                return node